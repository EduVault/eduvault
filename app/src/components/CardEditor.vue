<template>
  <div class="editor-wrapper u-scroller">
    <div class="editor">
      <p class="form__top-label">Front</p>
      <quill-editor
        class="flashcard__content m-2"
        ref="myQuillEditorFront"
        v-model="newFrontText"
        :options="editorOptions"
      ></quill-editor>
      <!-- <input v-model="newFrontText" name="card-front-input" class="form__text-input" type="text" /> -->
      <p class="form__top-label">Back</p>
      <!-- <input v-model="newBackText" name="card-back-input" class="form__text-input" type="text" /> -->
      <quill-editor
        class="flashcard__content m-2"
        ref="myQuillEditorBack"
        v-model="newBackText"
        :options="editorOptions"
      ></quill-editor>
      <div class="form__button-row ">
        <font-awesome-icon
          class="form__button button form__button--cancel"
          icon="times"
          size="2x"
          @click="$emit('closeEditor')"
        ></font-awesome-icon>
        <font-awesome-icon
          class="form__button button form__button--confirm primary"
          icon="check"
          size="2x"
          @click="newCard ? addCard() : editCard()"
        ></font-awesome-icon>
      </div>
      <div v-if="newCard">
        <div class="form__top-label">
          Add card to deck:
          <strong class="form__top-label--strong">{{ selectedDeck.title }}</strong>
        </div>
        <span v-show="decks.length > 1" class="tag-selection">
          <span class="tag-selection__title">Change deck:</span>
          <span v-for="deck in decks" :key="deck._id" class="tag-selection__tag-span">
            <button
              v-show="deck._id !== selectedDeck._id && !deck.deleted"
              class="tag-selection__tag"
              @click="$emit('changeSelectedDeck', deck._id)"
            >
              {{ deck.title }}
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
</template>

<script>
import { v4 as uuid } from 'uuid';
import { Quill, quillEditor } from 'vue-quill-editor';
import 'quill/dist/quill.snow.css';
import imageUpload from 'quill-plugin-image-upload';
import { uploadPictureToBucket } from '../store/textileHelpers';

Quill.register('modules/imageUpload', imageUpload);
const toolbarContent = [
  [{ header: [1, 2, 3, false] }],
  ['bold', 'italic', 'link', 'code-block', 'image'],
  ['clean'],
];

import store from '../store';
export default {
  components: { quillEditor },
  props: {
    editPayload: {
      type: Object,
      default: function() {
        return {
          deckId: {
            type: String,
            default: '',
          },
          frontText: {
            type: String,
            default: '',
          },
          backText: {
            type: String,
            default: '',
          },
          _id: {
            type: String,
            default: '',
          },
        };
      },
    },
    selectedDeck: {
      type: Object,
      default() {
        return {
          cards: [],
          title: 'none selected',
          _id: '',
          deleted: false,
          ttl: 1596161096048,
        };
      },
    },
    decks: {
      type: Array,
      default() {
        return [];
      },
    },
    newCard: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      newFrontText: '',
      newBackText: '',
      editorOptions: {
        theme: 'snow',
        modules: {
          imageUpload: {
            upload: async file => {
              return await uploadPictureToBucket(file);
            },
          },
          toolbar: toolbarContent,
          history: {
            delay: 2000,
            maxStack: 500,
            userOnly: true,
          },
        },
      },
    };
  },
  created() {
    this.setFields();
  },

  methods: {
    setFields: function() {
      if (!this.newCard) {
        this.newFrontText = JSON.parse(JSON.stringify(this.editPayload.frontText));
        this.newBackText = JSON.parse(JSON.stringify(this.editPayload.backText));
      }
    },
    addCard: function() {
      const payload = {
        card: {
          frontText: this.newFrontText,
          backText: this.newBackText,
          _id: uuid(),
          updatedAt: new Date().getTime(),
        },
        deckId: this.selectedDeck._id,
      };
      this.$emit('addCard', payload);
    },
    editCard: function() {
      const payload = {
        card: {
          frontText: this.newFrontText,
          backText: this.newBackText,
          _id: this.editPayload._id,
          updatedAt: new Date().getTime(),
        },
        deckId: this.editPayload.deckId,
      };
      this.$emit('editCard', payload);
    },
  },
};
</script>

<style scoped>
.editor-wrapper {
  z-index: 5000;
  overflow: scroll;
  position: fixed;
  top: 100px;
  padding: 20px;
  height: calc(100% - 150px);
  left: 0;
  display: flex;
  width: 100%;
  justify-content: center;
}

.editor {
  background-color: white;
  box-shadow: 0px 0px 14px 1px #0000008c;
  text-align: center;
  border-radius: 1px;
  overflow: scroll;
}
.tag-selection__tag {
  color: white;
  background: grey;
  border: none;
  border-radius: 7px;
  margin: 5px;
  padding: 0.35rem;
  cursor: pointer;
  font-weight: bold;
}
.flashcard__content >>> img {
  width: 100%;
  margin: auto;
  object-fit: fill;
}
.flashcard__content >>> .ql-align-center {
  text-align: center;
}
.flashcard__content >>> .ql-align-right {
  text-align: right;
}
.flashcard__content >>> .ql-align-left {
  text-align: left;
}
.flashcard__content >>> .ql-align-justify {
  text-align: justify;
}
.flashcard__content >>> p {
  font-size: 1em;
}
.flashcard__content >>> h3 {
  font-size: 1.5em;
}
.flashcard__content >>> h2 {
  font-size: 2em;
}
.flashcard__content >>> h1 {
  font-size: 3.5em;
}
.flashcard__content >>> em {
  font-style: italic;
}
</style>
