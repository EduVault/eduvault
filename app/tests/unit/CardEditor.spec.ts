import { mount, createLocalVue } from '@vue/test-utils';
import CardEditor from '@/components/CardEditor.vue';
import { library } from '@fortawesome/fontawesome-svg-core';
import {
  faTrashAlt,
  faPlusSquare,
  faCheck,
  faEdit,
  faTimes,
} from '@fortawesome/free-solid-svg-icons';
import {
  FontAwesomeIcon,
  //  FontAwesomeLayers
} from '@fortawesome/vue-fontawesome';
const localVue = createLocalVue();
library.add(faTrashAlt, faPlusSquare, faCheck, faEdit, faTimes);
localVue.component('fontAwesomeIcon', FontAwesomeIcon);

describe('CardEditor.vue', () => {
  let frontInput: any;
  let backInput: any;
  let cancelButton: any;
  let confirmButton: any;

  const wrapper = mount(CardEditor, {
    localVue,
    propsData: {
      editPayload: {
        deckId: 'MockID',
        frontText: 'MockFrontText',
        backText: 'MockBackText',
        _id: '12f14f23',
      },
    },
    data() {
      return {
        newFrontText: '',
        newBackText: '',
      };
    },
  }) as any;

  frontInput = wrapper.findAll('.card-editor .form__text-input').at(0);
  backInput = wrapper.findAll('.card-editor .form__text-input').at(1);
  cancelButton = wrapper.find('.card-editor .form__button--cancel');
  confirmButton = wrapper.find('.card-editor .form__button--confirm');
  it('renders', () => {
    expect(wrapper.exists()).toBe(true);
  });
  it('matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  it('updates changed text', () => {
    frontInput.setValue('value1');
    expect(wrapper.vm.newFrontText).toBe('value1');
    backInput.setValue('value2');
    expect(wrapper.vm.newBackText).toBe('value2');
  });

  it('outputs the correct data when confirm is clicked', async () => {
    frontInput.setValue('frontOutputValue');
    backInput.setValue('backOutputValue');
    await confirmButton.trigger('click');
    const expectedOutput = {
      frontText: 'frontOutputValue',
      backText: 'backOutputValue',
      deckId: 'MockID',
      _id: '12f14f23',
    };
    expect(JSON.stringify(wrapper.emitted().editCard[0][0])).toBe(JSON.stringify(expectedOutput));
  });
  it('emits cancelEdit when cancel button clicked', async () => {
    await cancelButton.trigger('click');
    expect(wrapper.emitted().cancelEdit).toBeTruthy();
  });
  it('has font awesome icon buttons', () => {
    const buttons = wrapper.findAllComponents(FontAwesomeIcon);
    // this works in mount but not shallowMount
    expect(buttons).toHaveLength(2);
  });
  it('has 2 icon buttons', () => {
    const buttons = wrapper.findAll('.card-editor .button');
    expect(buttons).toHaveLength(2);
  });
});
