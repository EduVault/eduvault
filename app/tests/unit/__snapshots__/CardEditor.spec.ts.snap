// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardEditor.vue matches snapshot 1`] = `
<div
  class="card-editor"
>
  <p
    class="form__top-label"
  >
    Front
  </p>
   
  <input
    class="form__text-input"
    type="text"
  />
   
  <p
    class="form__top-label"
  >
    Back
  </p>
   
  <input
    class="form__text-input"
    type="text"
  />
   
  <div
    class="form__button-row"
  >
    <svg
      aria-hidden="true"
      class="form__button button form__button--cancel svg-inline--fa fa-times fa-w-11 fa-2x"
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
     
    <svg
      aria-hidden="true"
      class="form__button button form__button--confirm svg-inline--fa fa-check fa-w-16 fa-2x"
      color="#42b983"
      data-icon="check"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
